src/backend/renderer/pixman/mod.rs

```diff
@@ -1216,26 +1204,25 @@ impl Bind<Dmabuf> for PixmanRenderer {
     }
 }
 
-impl Offscreen<PixmanRenderBuffer> for PixmanRenderer {
+impl Offscreen<Image<'static, 'static>> for PixmanRenderer {
     #[profiling::function]
     fn create_buffer(
         &mut self,
         format: DrmFourcc,
         size: Size<i32, BufferCoords>,
-    ) -> Result<PixmanRenderBuffer, <Self as Renderer>::Error> {
+    ) -> Result<Image<'static, 'static>, Self::Error> {
         let format_code =
             FormatCode::try_from(format).map_err(|_| PixmanError::UnsupportedPixelFormat(format))?;
         let image = pixman::Image::new(format_code, size.w as usize, size.h as usize, true)
             .map_err(|_| PixmanError::Unsupported)?;
-        Ok(PixmanRenderBuffer::from(image))
+        Ok(image)
     }
 }
 
-impl Bind<PixmanRenderBuffer> for PixmanRenderer {
-    #[profiling::function]
-    fn bind(&mut self, target: PixmanRenderBuffer) -> Result<(), <Self as Renderer>::Error> {
-        self.target = Some(PixmanTarget::RenderBuffer(target));
-        Ok(())
+impl Bind<Image<'static, 'static>> for PixmanRenderer {
```

Comment:
We should allow to bind an `Image` with non `'static` data lifetime. This would allow callers to bind `byte[]` slices in a safe way without having to rely on the ugly hack we do in the `DrmCompositor` for binding the dumb buffer slice.

Reply:
I couldn't find a way to introduce the necessary bound of `Image<'a, 'a>` outlives `PixmanTarget<'b>`. I want to add a `where 'a: 'b`-bound to `bind`, but that isn't compatible with the trait definition.

Adding `'b` to `Bind<'b>` doesn't help, as we still can't express the bound in the trait, `Target: 'b` doesn't help.

Nor does `where Target: 'a` even though that should imply, that `bind` should only be available if `Image<'b, 'b>` outlives `PixmanTarget<'a>`. The compiler doesn't seem smart enough.

Adding another lifetime to `PixmanTarget` makes it impossible to implement `RendererSuper`.

I couldn't find a way to specify the additional requirement with an higher-order bound (`for<'b>`), but that feels like what is necessary here.

Reply:
I see, that's quite unfortunate. I also unsuccessfully tried to find a safe way to express this.
Would be really great if we could use this to basically bind a scoped slice  :/

Theoretically is should just work as the `'bits` inside of the `Image` are already guaranteed to outlive any reference to the `Image`. In `PixmanTarget`/`PixmanTargetInternal` we only care that the `'bits` of the `Image` live at least as long as the reference of the image itself. So the only thing blocking this is that we can not prove it to the compiler. I wonder if it would be safe to just cast the lifetime in bind.

Reply:
Quick test of the change: https://github.com/cmeissl/smithay/commit/0e5eee5f59693907f87becfc71c6dba32db8e999